jetstream/analysis.py

```diff
@@ -157,209 +159,290 @@ def _calculate_metrics(
             )
 
         if analysis_length_dates < 0:
             logging.error(
                 "Proposed enrollment longer than analysis dates length:"
-                + f"{self.config.experiment.normandy_slug}"
+                + f"{config.experiment.normandy_slug}"
             )
             raise Exception("Cannot validate experiment")
 
         limits = TimeLimits.for_single_analysis_window(
             last_date_full_data=end_date.strftime("%Y-%m-%d"),
             analysis_start_days=0,
             analysis_length_dates=analysis_length_dates,
-            first_enrollment_date=self.config.experiment.start_date.strftime("%Y-%m-%d"),
+            first_enrollment_date=config.experiment.start_date.strftime("%Y-%m-%d"),
             num_dates_enrollment=dates_enrollment,
         )
 
         exp = mozanalysis.experiment.Experiment(
-            experiment_slug=self.config.experiment.normandy_slug,
-            start_date=self.config.experiment.start_date.strftime("%Y-%m-%d"),
+            experiment_slug=config.experiment.normandy_slug,
+            start_date=config.experiment.start_date.strftime("%Y-%m-%d"),
         )
 
         metrics = set()
-        for v in self.config.metrics.values():
+        for v in config.metrics.values():
             metrics |= {m.metric for m in v}
 
         sql = exp.build_query(
             metrics,
             limits,
             "normandy",
-            self.config.experiment.enrollment_query,
+            config.experiment.enrollment_query,
         )
 
         dry_run_query(sql)
 
-    def run(self, current_date: datetime, dry_run: bool = False) -> None:
+    def _save_statistics(
+        self,
+        period: AnalysisPeriod,
+        segment_results: List[Dict[str, Any]],
+        metrics_table: str,
+        normandy_slug: str,
+    ):
+        """Write statistics to BigQuery."""
+        job_config = bigquery.LoadJobConfig()
+        job_config.schema = StatisticResult.bq_schema
+        job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_TRUNCATE
+
+        # wait for the job to complete
+        self.bigquery.load_table_from_json(
+            segment_results, f"statistics_{metrics_table}", job_config=job_config
+        )
+
+        self._publish_view(period, normandy_slug=normandy_slug, table_prefix="statistics")
+
+    def run(
+        self, current_date: datetime, config: AnalysisConfiguration, dry_run: bool = False
+    ) -> None:
         """
         Run analysis using mozanalysis for a specific experiment.
         """
-        logger.info("Analysis.run invoked for experiment %s", self.config.experiment.normandy_slug)
+        logger.info("Analysis.run invoked for experiment %s", config.experiment.normandy_slug)
+
+        self.check_runnable(config, current_date)
+
+        # set up dask
+        cluster = LocalCluster(
+            dashboard_address="127.0.0.1:8782", processes=True, threads_per_worker=1
+        )
```

Comment:
Dask let's you expose a dashboard which is quite helpful for debugging.

We should also [avoid multi-threading and use processes instead](https://docs.dask.org/en/latest/scheduling.html#local-threads) since our code is dominated by Python code, otherwise we don't see any speedup due Python’s Global Interpreter Lock.

Reply:
The port number looks like something we should at least pull into a constant.